File: examples/telephone/mangler.rb
require 'pp'

class Mangler
  
  def self.mangle_word what
    patterns = [ what, what+"." ]
    index = 0
    (what.length).times do

      # delete a character
      if what.length > 3
        mod = what.clone
        mod[index] = ""
        patterns << mod
      end

      # insert a character
      mod = what.clone
      mod[index,0] = "."
      patterns << mod

      #change a character (biased towards making words longer)
      if what.length > 3
        mod = what.clone
        mod[index] = "."
        patterns << mod
      end

      index += 1
    end
    patterns = "^((" + patterns.join( ")|(" ) + "))$"
    # pp patterns
    matches = words.grep Regexp.new( patterns, Regexp::IGNORECASE )
    # pp matches
    matches[rand( matches.length )]
  end

  def self.mangle words
    words.split.map { |word| mangle_word word }.join(" ")
  end

  private

  def self.words
    @words ||=
      begin
        lines = IO.readlines File.join( File.dirname(__FILE__), "3esl.txt")
        lines.each { |line| line.chomp! } 
        lines
      end
  end

end

File: examples/telephone/mangler.py
import re
import random

import inspect
import sys
import os.path

class Mangler(object):
    
    @classmethod
    def mangle_word(cls,what):
        # print "what", what
        patterns = [ what, what + "." ]
        for index in xrange(len(what)):

            if len(what) > 3:
                mod = what
                mod = mod[0:index]+mod[index+1:]
                patterns.append(mod)

            mod = what
            mod = mod[0:index] + "." + mod[index:]
            patterns.append(mod)

            if len(what) > 3:
                mod = what
                mod = mod[0:index] + "." + mod[index+1:]
                patterns.append(mod)

        # print repr(patterns)

        patterns = "^((" + ")|(".join(patterns) + "))$"
        regexp = re.compile( patterns, re.IGNORECASE )
        matches = [ word for word in cls._words() if regexp.match( word ) ]
        # print "a", repr(matches)
        # print "b", repr(patterns)
        # print "c", repr(matches)
        return matches[random.randint( 0, len(matches)-1 )]

    @classmethod
    def mangle(cls,words):
        return " ".join( [ cls.mangle_word( word ) for word in words.split() ] )

    @classmethod
    def _words(cls):
        try:
            return cls._word_list
        except:
            f = os.path.join( os.path.dirname( inspect.getabsfile( inspect.currentframe() ) ), "3esl.txt" )
            f = open(f)
            cls._word_list = [ l[0:-1] for l in f ]
            return cls._word_list
